File: Residential_Controller.py
# Residential Elevator Controller In Python From My Pseudocode From Last Week
# Written by : Valerie Beaupre
# Date : 2018/09/24 TO 2018/09/27

import itertools
import os
import time
import datetime
from datetime import timedelta

nbColumn = 1
nbElev = 2
nbFloors = 10
floorNames = ["BSMT 1", "BSMT 2", "Lobby", "2nd","3rd", "4th", "5th", "6th", "7th", "8th"]
lobby = 3
defaultFloor = 3
timePerFloor = 1000

# Elevator Directions
directionNameList = ["GoingNowhere", "Down", "Up"]
GoingNowhere = 0
Down = 1
Up = 2

# People entering or leaving the elevator
inOrOutNameList = ["GoingIn","GoingOut"]
GoingIn = 0
GoingOut = 1

#Elevator status
elevatorStatusNameList = ["Idle","Stopping","Stopped","Moving"]
Idle = 0
Stopping = 1
Stopped = 2
Moving = 3

#Door status
oorStatusNameList= ["Closed","Closing","Opening","Opened"]
Closed = 0
Closing = 1
Opening = 2
Opened = 3

#Buttons Function
AddFloor = 0
CallElevator = 1
OpenDoor = 2
CloseDoor = 3

#Buttons status
Inactive = 0
Active = 1

delayElevatorStopping = 2000
delayDoorOpening = 1000
delayBeforeCloseDoor = 5000
delayForceCloseDoor = 50000
timeOutDoorOpen = 15000
delayMaxIdleTime = 15000   
currentFloor = 0
appTimeOut = 60000

timeInMilli = lambda: int(round(time.time() * 1000))

def counter(): #Self increment counter to generate unique ID's
    return lambda c=itertools.count(): next(c)

count = counter()  


class ElevatorController:
    def __init__(self):
        self.id = 1
        self.columns = Column(2, lobby, 2)
        print("Controller Created a column for " + str(nbElev) + " elevators and for " + str(nbFloors) + " floors")

    def requestElevator(self, floorNumber, direction):
        print("Elevator requested at " + floorNames[floorNumber-1] + " to go " + directionNameList[direction])
        elevator = self.callElevator(direction, floorNumber, GoingIn)
        return elevator

    def requestFloor(self, elevator, requestedFloor):
        self.addDestinationElev(requestedFloor, elevator, GoingOut,"")

    def callElevator(self, requestedDirection, requestedFloor, isGoingInOrOut):
        elevator = self.findElevator(requestedDirection, requestedFloor)
        if self.checkIfDestinExist(requestedFloor, elevator) is False:
            self.addDestinationElev(requestedFloor, elevator, isGoingInOrOut, requestedDirection)
        return elevator

    def checkIfDestinExist(self, floor, elevator):
        if elevator.destinationList:
            for destination in elevator.destinationList:
                if destination == floor:
                    return True
        return False

    def findElevator(self, requestedDirection, requestedFloor):
        idleElevatorList = []
        for elevator in self.columns.elevators:
            if elevator.door.alarm is False:
                if requestedFloor is elevator.currentFloor:
                    if (elevator.status is Stopped and elevator.direction is requestedDirection) or (elevator.status is Idle and not elevator.destinationList):
                        return elevator
                elif ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Up) or (elevator.direction is GoingNowhere)) and (requestedDirection is Up) and ((elevator.status is Moving) or (elevator.status is Stopped))):
                    return elevator
                elif ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Down) or (elevator.direction is GoingNowhere)) and (requestedDirection is Down) and ((elevator.status is Moving) or (elevator.status is Stopped))):
                    return elevator
                elif elevator.status is Idle and not elevator.destinationList:
                    idleElevatorList.append(elevator)
        if idleElevatorList:
            if (len(idleElevatorList) > 1):
                gap = 99999
                for elevator in idleElevatorList:
                    if gap > abs(elevator.currentFloor - requestedFloor):
                        gap = abs(elevator.currentFloor - requestedFloor)
                        elevatorToUse = elevator
                return elevatorToUse
            else:
                return idleElevatorList[0]

        elevator = self.nearestElevator(requestedFloor, requestedDirection)
        if not elevator:
            return elevator
        else:
            return self.shortestdestinationList()

    def shortestdestinationList(self):
        length = 99999
        #elevWithShortestList = None
        for elevator in self.columns.elevators:
            if length > len(elevator.destinationList):
                length = len(elevator.destinationList)
                elevWithShortestList = elevator
        return elevWithShortestList
        print("shortestList")

    def nearestElevator(self, requestedFloor, requestedDirection):
        gap = 99999
        elevWithShortestGap = None
        for elevator in self.columns.elevators:
            if (gap > abs(elevator.currentFloor - elevator.destinationList[0])) and (elevator.door.alarm is False):
                if ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Up) or (elevator.direction is GoingNowhere)) and (requestedDirection is Up)) or ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Down) or (elevator.direction is GoingNowhere)) and (requestedDirection is Down)):
                    gap = abs(elevator.currentFloor - elevator.destinationList[0])
                    elevWithShortestGap = elevator
        return elevWithShortestGap
        print("nearestELev")

    def addDestinationElev(self, floor, elevator, isGoingInOrOut, requestedDirection):
        if elevator.destinationList:
            for destination in elevator.destinationList:
                if elevator.direction:
                    if elevator.direction is Up:
                        if (floor < destination) and (floor > elevator.currentFloor):
                            insertPosition = elevator.destinationList.index(destination)
                            elevator.destinationList.insert(insertPosition, floor)
                            elevator.inOutList.insert(insertPosition, isGoingInOrOut)
                            elevator.directionList.insert(insertPosition, requestedDirection)
                            return
                        elif (floor > destination) and (floor < elevator.currentFloor):
                            insertPosition = elevator.destinationList.index(destination)
                            elevator.destinationList.insert(insertPosition, floor)
                            elevator.inOutList.insert(insertPosition, isGoingInOrOut)
                            elevator.directionList.insert(insertPosition, requestedDirection)
                            return
                    elif elevator.direction is Down:
                        if (floor > destination) and (floor < elevator.currentFloor):
                            insertPosition = elevator.destinationList.index(destination)
                            elevator.destinationList.insert(insertPosition, floor)
                            elevator.inOutList.insert(insertPosition, isGoingInOrOut)
                            elevator.directionList.insert(insertPosition, requestedDirection)
                            return
                        elif (floor < destination) and (floor > elevator.currentFloor):
                            insertPosition = elevator.destinationList.index(destination)
                            elevator.destinationList.insert(insertPosition, floor)
                            elevator.inOutList.insert(insertPosition, isGoingInOrOut)
                            elevator.directionList.insert(insertPosition, requestedDirection)
                            return

                else:
                    elevator.destinationList.append(floor)
                    elevator.inOutList.append(isGoingInOrOut)
                    elevator.directionList.append(requestedDirection)
                    return
        else:
            elevator.destinationList.append(floor)
            elevator.inOutList.append(isGoingInOrOut)
            elevator.directionList.append(requestedDirection)
            return

    def checkElevatorStatus(self):
        for elevator in self.columns.elevators:
            if (timeInMilli() > elevator.openDoorTime + timeOutDoorOpen) and (elevator.door.status is not Closed):
                elevator.door.alarm = True
                elevator.forceCloseDoor()
            if (timeInMilli() > elevator.idleTime + delayMaxIdleTime) and (not elevator.destinationList):
                pass
                #self.AddDestination(self.Columns.DefaultFloor,Elevator,GoingOut,"")

            if (timeInMilli() > (elevator.openDoorTime + delayDoorOpening)) and (elevator.door.status is Opening):
                elevator.door.status = Opened
                print("Elevator " + str(elevator.id) + " door is opened") 
            
            if ((timeInMilli() > (elevator.openDoorTime + delayBeforeCloseDoor)) and (elevator.door.status is Opened)) or elevator.door.alarm is True:
                elevator.closeDoor()
            
            if (timeInMilli() > elevator.closeDoorTime + delayDoorOpening) and elevator.door.status is Closing:
                if (elevator.door.alarm is True) and (elevator.door.status is not Opening):
                    elevator.openDoor()
                    print("Elevator " + str(elevator.id) + " door is opening again because of obstruction")
                elevator.door.status = Closed
                elevator.status = Idle
                elevator.idleTime = timeInMilli()
                #print("CloseDoor 1")
                print("Elevator " + str(elevator.id) + " door is closed")
            
            if (timeInMilli() > elevator.forceCloseTime + delayForceCloseDoor) and (elevator.door.status is Closing) and elevator.door.alarm is False:
                elevator.status = Idle
                elevator.idleTime = timeInMilli()
                elevator.alarm = False
                elevator.door.alarm = False
                elevator.door.status = Closed
                #print("CloseDoor 2")
                print("Elevator " + str(elevator.id) + " door is closed")

            if (timeInMilli() > elevator.stopTime + delayElevatorStopping) and elevator.status is Stopping:
                elevator.Status = Stopped
                elevator.Direction = GoingNowhere
                print("Elevator " + str(elevator.id) + " is stopped at " + str(floorNames[elevator.currentFloor-1]) + " floor for people " + inOrOutNameList[elevator.inOutList[0]]) 
                elevator.directionList.pop(0)
                elevator.inOutList.pop(0)

    def verifyDestinationList(self):
        for elevator in self.columns.elevators:
            if elevator.destinationList:
                if elevator.inOutList:
                    if (elevator.destinationList[0]) and (elevator.status is Idle):
                        if (elevator.currentFloor is not elevator.destinationList[0]) and elevator.door.status is Closed:
                            elevator.startMove()
                        elif (elevator.inOutList[0] is GoingIn) and (elevator.currentFloor is elevator.destinationList[0]) and elevator.door.status is Closed:
                            print("Open Door 1")
                            elevator.destinationList.pop(0)
                            elevator.directionList.pop(0)
                            elevator.inOutList.pop(0)
                            elevator.openDoor()

    def checkMovingElevator(self):
        for elevator in self.columns.elevators:
            if (elevator.status is Moving) and (elevator.door.alarm is False):
                if (timeInMilli() - elevator.moveTimeStamp) >= timePerFloor:
                    if elevator.currentFloor < elevator.destinationList[0]:
                        elevator.floorLevelForTiming += 1
                    else:
                        elevator.floorLevelForTiming -= 1

                    elevator.moveTimeStamp = timeInMilli()
                    if elevator.currentFloor is not elevator.floorLevelForTiming:
                        elevator.currentFloor = elevator.floorLevelForTiming
                        print("Elevator " + str(elevator.id) + " is at " + floorNames[elevator.currentFloor-1] + " floor")
                        
                    if elevator.currentFloor is elevator.destinationList[0]:
                        elevator.stopElevator()
                        self.clearButtons(elevator)
                        elevator.destinationList.pop(0)
                        #print(elevator.destinationList)

            if elevator.status is Stopped and elevator.door.status is Closed:
                #print("OpenDoor 2")
                elevator.openDoor()

    def clearButtons(self, elevator):
        if elevator.directionList[0] is not "":
            button = self.columns.findDirectionButton(elevator.directionList[0],elevator.destinationList[0]) 
            button.status = Inactive
            print("Request button direction " + directionNameList[button.direction] + " at " + floorNames[elevator.destinationList[0]-1] + " floor is inactive")
        button = elevator.findFloorButton()
        button.status = Inactive
        print("Elevator " + str(elevator.id) + " " + floorNames[elevator.currentFloor-1] + " floor button is inactive")


class Floor:
    def __init__(self, id, name, buttons):
        self.id = id
        self.name = name
        #self.directionButtons = buttons


class Column:
    def __init__(self, nbElev, defaultFloor, nbDirectionButtons):
        self.nbElev = nbElev
        self.defaultFloor = defaultFloor
        self.nbDirectionButtons = nbDirectionButtons
        self.elevators = []
        for index in range(self.nbElev):
            self.elevators.append(Elevator(index+1))
        #for elevator in range(self.elevators):
            #print("elevator.id")

        self.destinationList = []
        self.directionButtons = []
        for index in range(nbFloors):
            directionButtonList = []
            if index == 0:
                directionButtonList.append(DirectionButton(Up))
            elif index == (nbFloors - 1):
                directionButtonList.append(DirectionButton(Down))
            else:
                directionButtonList.append(DirectionButton(Up))
                directionButtonList.append(DirectionButton(Down))

            self.destinationList.append(Floor(index, floorNames[index], directionButtonList))

    def findDirectionButton(self, direction, requestedFloor):
        for button in self.directionButtons:
            if (requestedFloor == self.directionButtons) and (direction == self.directionButtons):
                return button
        print("Column Created")


class Elevator:
    def __init__(self, id):
        self.id = id
        self.currentFloor = lobby
        self.direction = GoingNowhere
        self.status = Idle
        self.floorLevelForTiming = lobby
        self.alarm = False
        self.moveTimeStamp = 0
        self.idleTime = timeInMilli()
        self.openDoorTime = timeInMilli()
        self.closeDoorTime = timeInMilli()
        self.stopTime = timeInMilli()
        self.forceCloseTime = timeInMilli()
        self.destinationList = []     
        self.inOutList = []
        self.directionList = []
        self.floorButtons = []  #Buttons inside the elevators
        for index in range(nbFloors):
            self.floorButtons.append(FloorButton(index))

        self.door = Door()
        self.openDoorButton = OpenDoorButton(self.door)
        self.closeDoorButton = CloseDoorButton(self.door)
        print("Elevator Created")

    def findFloorButton (self, currentFloor):
        for button in self.floorButtons: 
            if self.currentFloor is self.floorButtons:
                return button 

    def startMove(self):
        if (self.currentFloor < self.destinationList[0]): 
            self.direction : Up
        elif (self.currentFloor > self.destinationList[0]): 
            self.direction : Down
        self.status : Moving
        self.moveTimeStamp = timeInMilli()
        self.floorLevelForTiming = self.currentFloor
        #print("Elevator " + str(self.id) + " is moving " + directionNameList[self.direction] + " to " + floorNames[self.destination[0]-1] + " floor for people " + inOrOutNameList[self.inOutList[0]])
        print("Elevator " + str(self.id) + " is moving " + directionNameList[self.direction] + " to " + floorNames[self.destinationList[0]-1] + " floor")

    def stopElevator(self):
        self.stopTime = timeInMilli()
        self.Status = Stopping
        print("Elevator " + str(self.id) + " is stopping at " + floorNames[self.destinationList[0]-1] + " floor")

    def openDoor(self):
        self.door.status = Opening
        self.OpenDoorTime = timeInMilli()
        print("Elevator " + str(self.id) + " door is opening")

    def closeDoor(self):
        self.door.status = Closing
        self.closeDoorTime = timeInMilli()
        print("Elevator " + str(self.id) + " door is closing")

    def forceCloseDoor(self):
        self.alarm = True 
        print("Elevator " + str(self.id) + " door is losing slowly (Force Close)")
        self.forceCloseTime = timeInMilli()


class DirectionButton:
    def __init__(self, direction):
        self.direction = ""  # UP or DOWN
        self.status = Inactive  # Active or Inactive


class FloorButton:
    def __init__(self, floor):
        self.floor = floor
        self.status = Inactive


class Door:
    def __init__(self):
        self.id = id
        self.status = Closed
        self.obstruction = False
        self.alarm = False


class OpenDoorButton:
    def __init__(self, door):
        self.door = door


class CloseDoorButton:
    def __init__(self, door):
        self.door = door


controller = ElevatorController()

def initTest(elev1Floor,elev2Floor):
    controller.columns.elevators[0].currentFloor = elev1Floor
    controller.columns.elevators[0].floorLevelForTiming = elev1Floor
    print("Elevator 1 current floor is " + floorNames[controller.columns.elevators[0].currentFloor-1])
    
    controller.columns.elevators[1].currentFloor = elev2Floor
    controller.columns.elevators[1].floorLevelForTiming = elev2Floor
    print("Elevator 2 current floor is " + floorNames[controller.columns.elevators[1].currentFloor-1])


"""
FLOOR INDEX:
    BSMT 2 : 1
    BSMT 1 : 2
    Lobby : 3
    2nd : 4
    3rd : 5
    4th : 6
    5th : 7
    6th : 8
    7th : 9
    8th : 10

ELEVATOR INDEX:
    1 : 1
    2 : 2

STATUS INDEX:
    GOINGNOWHERE : 0
    DOWN : 1
    UP : 2

"""

#CONTROLLER INITIALISATION
initTest(10,3)
controllerRunning = True

def main():
    elevator = []
    firstInstructionDone = False
    timeOut = timeInMilli()
    readyToStopController = False 

    while controllerRunning is True:
        #INITIATE ALL THE CALLS
        if firstInstructionDone is False:
            firstInstructionDone = True
            #Firt user
            elevator.append(controller.requestElevator(10, Down))
            #Second user
            elevator.append(controller.requestElevator(1, Up))
            #Third user
            #Elevator.append(controller.requestElevator(9, Down))
            #First user
            controller.requestFloor(elevator[0],3)
            #Second user
            controller.requestFloor(elevator[1],5)
            #Third user
            #controller.requestFloor(elevator[2],1)
    

        #CONTROLLER SEQUENCES
        controller.checkElevatorStatus()
        controller.verifyDestinationList()
        controller.checkMovingElevator()


        #STOP THE CONTROLLER WHEN DONE OR TIMED OUT
        if ((timeInMilli() > controller.columns.elevators[0].idleTime + appTimeOut) and (timeInMilli() > controller.columns.elevators[1].idleTime + appTimeOut)):
            readyToStopController = True
        if (timeInMilli() > timeOut + appTimeOut) or (readyToStopController is True):
            print("Break")
            break

main()


# **********************************************************************************
# **********************************************************************************

""" ****************************************************************************************

Residential Scenario 1:
A User located at Floor 1 calls for elevators Originating from Basement 1 and Floor 4, he gets one and gets to the Fifth floor with it'

Residential Scenario 2:
A User located at Floor Basement 2 calls for elevators Originating from Floor 8 and Floor 1, he gets one to get to the 4th floor, Simultaneously, someone at Floor 1 requests an elevator to get to 3rd floor as someone at 7th requests to go down to basement 1'

Residential Scenario 3:
A User located at Floor 8 calls for elevators Originating from Floor 8 and Floor 1, he gets one to get to the 1st floor. Simultaneously Someone at Floor 1 requests an elevator to get to Basement 1'

 """

File: Residential_Controller.go
package main

import (
	"math"
	"fmt"
	"strconv"
	"strings"
	"time"
)

var floorName = []string{"BSMT 2", "BSMT 1", "Lobby", "2nd", "3rd", "4th", "5th", "6th", "7th", "8th"}
var directionNameList = []string{"GoingNowhere", "Down", "Up"}
var inOutNameList = []string{"GoingIn", "GoingOut"}
var doorStatusNameList = []string{"Idle", "Stopping", "Stopped", "Moving"}
var elevatorStatusNameList = []string{"Closed", "Closing", "Opening", "Opened"}

//Constants definition
const (
	nbFloors              int   = 10
	lobby                 int   = 3
	timePerFloor          int64 = 1000
	delayElevatorStopping int64 = 2000
	delayDoorOpening      int64 = 1000
	delayBeforeCloseDoor  int64 = 5000
	delayForceClose       int64 = 50000
	timeoutDoorOpen       int64 = 15000
	delayMaxIdleTime      int64 = 15000
	appTimeout            int64 = 60000

	//Elevator directions
	goingNowhere int = 0
	down         int = 1
	up           int = 2

	//People entering or leaving the elevator
	goingIn  int = 0
	gointOut int = 1

	//Elevator status
	idle     int = 0
	stopping int = 1
	stopped  int = 2
	moving   int = 3

	//Door status
	closed  int = 0
	closing int = 1
	opening int = 2
	opened  int = 3

	//Buttons function
	addFloor     int = 0
	callElevator int = 1
	openDoor     int = 2
	closeDoor    int = 3

	//Buttons status
	inactive int = 0
	active   int = 1
)

func getTimeInMilli() int64 {
	now := time.Now()
	unixNano := now.UnixNano()
	umillisec := unixNano / 1000000
	return umillisec
}

func join(strs ...string) string {
	var sb strings.Builder
	for _, str := range strs {
		sb.WriteString(str)
	}
	return sb.String()
}

// Counter :
type Counter struct {
	count int
}

func (c *Counter) currentValue() int {
	return c.count
}

func (c *Counter) increment() {
	c.count++
}

func insertInList(list []int, insertPos int, value int) []int {
	var listTemp []int
	for index := len(list) - 1; index >= insertPos-1; index-- {
		listTemp = append(listTemp, list[index])
		list = append(list[:index], list[index+1:]...)
	}
	list = append(list, value)

	for index := len(listTemp) - 1; index >= 0; index-- {
		list = append(list, listTemp[index])
	}

	return list
}

func findIndexInList(list []int, value int) int {
	for index := 0; index <= len(list); index++ {
		if list[index] == value {
			return index
		}
	}
	return -1
}

// OpenDoorButton :
type OpenDoorButton struct {
	door int
}

// CloseDoorButton :
type CloseDoorButton struct {
	door int
}

// FloorButton :
type FloorButton struct {
	floor  int
	status int
}

// DirectionButton :
type DirectionButton struct {
	floor     int
	direction int
	status    int
}

// Door :
type Door struct {
	ID          int
	status      int
	alarm       bool
	obstruction bool
}

// Elevator :
type Elevator struct {
	ID                  int
	currentFloor        int
	direction           int
	status              int
	floorLevelForTiming int
	alarm               bool
	moveTimeStamp       int64
	idleTime            int64
	openDoorTime        int64
	closeDoorTime       int64
	stopTime            int64
	forceCloseTime      int64
	destinationList     []int
	inOutList           []int
	directionList       []int
	floorButtons        []int
	door                *Door
	openDoorButton      *OpenDoorButton
	closeDoorButton     *CloseDoorButton
}

// Column :
type Column struct {
	nbElev              int
	defaultFloor        int
	nbDirectionButtons  int
	elevator            []int
	destinationList     []int
	directionButtonList []int
}

// Floor :
type Floor struct {
	ID   int
	name int
}

// ElevatorController :
type ElevatorController struct {
	ID      int
	columns *Column
}

func (elev *Elevator) stopElevator() {
	elev.stopTime = getTimeInMilli()
	elev.status = stopping
	fmt.Println(join("Elevator ", strconv.Itoa(elev.ID), " is stopping at ", floorName[elev.destinationList[0]-1], " floor"))
}

func (elev *Elevator) openDoor() {
	elev.door.status = opening
	elev.openDoorTime = getTimeInMilli()
	fmt.Println(join("Elevator ", strconv.Itoa(elev.ID), " door is opening"))
}

func (elev *Elevator) closeDoor() {
	elev.door.status = closing
	elev.closeDoorTime = getTimeInMilli()
	fmt.Println(join("Elevator ", strconv.Itoa(elev.ID), " door is closing"))
}

func (elev *Elevator) forceCloseDoor() {
	elev.alarm = true
	elev.forceCloseTime = getTimeInMilli()
	fmt.Println(join("Elevator ", strconv.Itoa(elev.ID), " door is closing slowly (Force close)"))
}

func (elev *Elevator) findFloorButton() {
	for _, button := range elev.floorButtons {
		if button.floor == elev.currentFloor {
			return button
		}
	}
}

func (elev *Elevator) startMove() {
	if elev.currentFloor < lev.destinationList[0] {
		elev.direction = up
	} else {
		elev.direction = down
	}
	elev.status = moving
	elev.floorLevelForTiming = elev.currentFloor
	fmt.Println(join("Elevator ", strconv.Itoa(elev.ID), " is moving ", directionNameList[elev.direction], " to ", floorName[elev.destinationList[0]-1], " floor"))
}

func (col *Column) findDirectionButton() {
	for _, button := range col.directionButtonList {
		if button.direction == requestedDirection && button.floor == requestedFloor {
			return button
		}
	}
	return nil
}

func (ctrl *ElevatorController) requestElevator(floorNumber int, direction int) *Elevator {
	fmt.Println(join("Elevator requested at ", floorNames[floorNumber-1], " to go ", directionNameList[direction]))
	elevator := ctrl.callElevator(direction, floorNumber, goingIn)
	return elevator
}

func (ctrl *ElevatorController) requestFloor(elevator, requestedFloor) {
	ctrl.addDestinationElev(requestedFloor, elevator, gointOut, -1)
	return
}

func (ctrl *ElevatorController) callElevator(requestedDirection int, requestedFloor int, isGoingInOrOut int) {
	elevator := ctrl.findElevator(requestedDirection, requestedFloor)
	if ctrl.checkIfDestinExist(requestedFloor, elevator) == false {
		ctrl.addDestinationElev(requestedFloor, elevator, gointOut, requestedDirection)
	}
	return
}

func (ctrl *ElevatorController) checkIfDestinExist(floor int, elevator *Elevator) bool {
	if len(elevator.destinationList) > 0 {
		for _, destination := range elevator.destinationList {
			if destination == floor {
				return true
			}
		}
	}
	return false
}

func (ctrl *ElevatorController) shortestdestinationList() *Elevator {
	length := 99999
	var shortestList *Elevator
	for _, elevator := ctrl.columns.elevators {
		if length > len(elevator.destinationList) {
			length = len(elevator.destinationList)
			elevWithShortestList = elevator
		}
	}
	return elevWithShortestList
}

func (ctrl *ElevatorController) nearestElevator(requestedFloor int, requestedDirection int) *Elevator {
	gap := 99999 
	var shortestGap *Elevator
	for _, elevator := range ctrl.columns.elevators {
		if (gap > int(math.Abs(float64(elevator.currentFloor - elevator.destinationList[0])))) && elevator.door.alarm == false {
			if ((requestedFloor > elevator.currentFloor) && ((elevator.direction == up) || (elevator.direction == goingNowhere)) && (requestedDirection == up)) || ((requestedFloor > elevator.direction == down || (elevator.direction == goingNowhere)) && (requestedDirection == down)) {
				gap = int(math.Abs(float64(elevator.currentFloor - elevator.destinationList[0])))
				elevWithShortestGap = elevator
			}
		}
	}
	return elevWithShortestGap
}

func (ctrl *ElevatorController) clearButtons(elevator *ElevatorController) {
	if elevator.directionList[0] > 0 {
		button := ctrl.columns.findDirectionButton(elevator.directionList[0], elevator.destinationList[0])
		button.status = inactive
		fmt.Println(join("Request button direction ", floorName[elevator.currentFloor-1], " floor is inactive"))
	}
	button := elevator.findFloorButton(elevator.destinationList[0])
	button.status = inactive
	fmt.Println(join("Elevator ", strconv.Itoa(elevator.ID), " ", floorName[elevator.currentFloor-1], " floor button is inactive"))
	return
}

func (ctrl *ElevatorController) checkMovingElevator() {
	for _, elevator := range ctrl.columns.elevators {
		if elevator.status == moving && elevator.door.alarm == false {
			if (getTimeInMilli() - elevator.moveTimeStamp) >= timePerFloor {
				if elevator.currentFloor < elevator.destinationList[0] {
					elevator.floorLevelForTiming++
				} else {
					elevator.floorLevelForTiming--
				}
				elevator.moveTimeStamp = getTimeInMilli
				if elevator.currentFloor != elevator.floorLevelForTiming {
					elevator.currentFloor = elevator.floorLevelForTiming
					fmt.Println(join("Elevator ", strconv.Itoa(elevator.ID), " is at ", floorName[elevator.currentFloor-1], "floor"))
				}
				if elevator.currentFloor == elevator.destinationList[0] {
					elevator.stopElevator()
					ctrl.clearButtons(elevator)
					elevator.destinationList = append(elevator.destinationList[:0], elevator.destinationList[1:]...)
				}
			}
		}
		if elevator.status == stopped && elevator.door.status == closed {
			elevator.openDoor()
		}
	}
	return
}

func (ctrl *ElevatorController) verifyDestinationList() {
	for _, elevator := range ctrl.columns.elevators {
		if (len(elevator.destinationList) > 0) && (len(elevator.inOutList) >0) {
			if elevator.destinationList[0] != 0 && elevator.status == idle {
				if elevator.currentFloor != elevator.destinationList[0] && elevator.door.status == closed {
					elevator.startMove()
				} else if (elevator.inOutList[0] && elevator.door.status == closed)
				elevator.destinationList = append(elevator.destinationList[:0], elevator.destinationList[1:]...)
				elevator.directionList = append(elevator.directionList[:0], elevator.directionList[1:]...)
				elevator.inOutList = append(elevator.inOutList[:0], elevator.inOutList[1:]...)
				elevator.openDoor()
			}
		}
	}
	return
}

func (ctrl *ElevatorController) checkElevatorStatus()

/*
def checkElevatorStatus(self):
	for elevator in self.columns.elevators:
		if (timeInMilli() > elevator.openDoorTime + timeOutDoorOpen) and (elevator.door.status is not Closed):
			elevator.door.alarm = True
			elevator.forceCloseDoor()
		if (timeInMilli() > elevator.idleTime + delayMaxIdleTime) and (not elevator.destinationList):
			pass
			#self.AddDestination(self.Columns.DefaultFloor,Elevator,GoingOut,"")

		if (timeInMilli() > (elevator.openDoorTime + delayDoorOpening)) and (elevator.door.status is Opening):
			elevator.door.status = Opened
			print("Elevator " + str(elevator.ID) + " door is opened") 
		
		if ((timeInMilli() > (elevator.openDoorTime + delayBeforeCloseDoor)) and (elevator.door.status is Opened)) or elevator.door.alarm is True:
			elevator.closeDoor()
		
		if (timeInMilli() > elevator.closeDoorTime + delayDoorOpening) and elevator.door.status is Closing:
			if (elevator.door.alarm is True) and (elevator.door.status is not Opening):
				elevator.openDoor()
				print("Elevator " + str(elevator.ID) + " door is opening again because of obstruction")
			elevator.door.status = Closed
			elevator.status = Idle
			elevator.idleTime = timeInMilli()
			print("Elevator " + str(elevator.ID) + " door is closed")
		
		if (timeInMilli() > elevator.forceCloseTime + delayForceCloseDoor) and (elevator.door.status is Closing) and elevator.door.alarm is False:
			elevator.status = Idle
			elevator.idleTime = timeInMilli()
			elevator.alarm = False
			elevator.door.alarm = False
			elevator.door.status = Closed
			print("Elevator " + str(elevator.ID) + " door is closed")

		if (timeInMilli() > elevator.stopTime + delayElevatorStopping) and elevator.status is Stopping:
			elevator.status = Stopped
			elevator.direction = GoingNowhere
			print("Elevator " + str(elevator.ID) + " is stopped at " + str(floorNames[elevator.currentFloor-1]) + " floor for people " + inOrOutNameList[elevator.inOutList[0]]) 
			elevator.directionList.pop(0)
			elevator.inOutList.pop(0)

*/

func (ctrl *ElevatorController) addDestinationElev(floor int, elevator *Elevator, isGoingInOrOut int, requestedDirection int) {
	if len(elevator.destinationList) > 0 {
		for _, destination := range elevator.destinationList {
			if elevator.direction != 0 {
				if elevator.direction == up {
					if (floor < destination) && (floor > elevator.currentFloor) {
						// insertPosition := elevator.destinationList = append(elevator.destinationList[:0], append([]T{x}, a[1:]...)...)
						insertPosition := findIndexInList(elevator.destinationList, destination)
						elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
						elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
						elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
						return
					} else if (floor > destination) && (floor < elevator.currentFloor) {
						insertPosition := findIndexInList(elevator.destinationList, destination)
						elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
						elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
						elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
						return
					}
				} else if elevator.direction == down {
					if (floor > destination) && (floor < elevator.currentFloor) {
						insertPosition := findIndexInList(elevator.destinationList, destination)
						elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
						elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
						elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
						return
					} else if (floor < destination) && (floor > elevator.currentFloor) {
						insertPosition := findIndexInList(elevator.destinationList, destination)
						elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
						elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
						elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
						return
					}
				}
			} else {
				elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
				elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
				elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
				return
			}
		}
	} else {
		elevator.destinationList = insertInList(elevator.destinationList, insertPosition, floor)
		elevator.inOutList = insertInList(elevator.inOutList, insertPosition, isGoingInOrOut)
		elevator.directionList = insertInList(elevator.directionList, insertPosition, requestedDirection)
		return
	}
	return
}



/*
*/

/*
def findElevator(self, requestedDirection, requestedFloor):
	idleElevatorList = []
	for elevator in self.columns.elevators:
		if elevator.door.alarm is False:
			if requestedFloor is elevator.currentFloor:
				if (elevator.status is Stopped and elevator.direction is requestedDirection) or (elevator.status is Idle and not elevator.destinationList):
					return elevator
			elif ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Up) or (elevator.direction is GoingNowhere)) and (requestedDirection is Up) and ((elevator.status is Moving) or (elevator.status is Stopped))):
				return elevator
			elif ((requestedFloor > elevator.currentFloor) and ((elevator.direction is Down) or (elevator.direction is GoingNowhere)) and (requestedDirection is Down) and ((elevator.status is Moving) or (elevator.status is Stopped))):
				return elevator
			elif elevator.status is Idle and not elevator.destinationList:
				idleElevatorList.append(elevator)
	if idleElevatorList:
		if (len(idleElevatorList) > 1):
			gap = 99999
			for elevator in idleElevatorList:
				if gap > abs(elevator.currentFloor - requestedFloor):
					gap = abs(elevator.currentFloor - requestedFloor)
					elevatorToUse = elevator
			return elevatorToUse
		else:
			return idleElevatorList[0]

	elevator = self.nearestElevator(requestedFloor, requestedDirection)
	if not elevator:
		return elevator
	else:
		return self.shortestdestinationList()
*/
